### PR TITLE
Chaospy-support for multivariate distributions

### PR DESCRIPTION
**Pull request by jonathf**
Extend Chaospy distribution support to include stochastically independent multivariate distributions.

### CODE REVIEW

**Comment by jonathf** [comment]
@jschueller, yes, thank you. I will add tests before submitting the PR for review.

Side note, the doctest example is commented out. Can I assume this is because doctests are preferred to be run without depending on chaospy being installed? If so, is there a reason the formulation isn't on the more canonical format:

    >>> import chaospy as cp  # doctest: +SKIP

**Comment by jschueller** [comment]
Yes, we could use that syntax, thanks for pointing that out

**Comment by jonathf** [comment]
@regislebrun,

I forgot to thank you for adding support for chaospy to OpenTURNS. So thank you. It is appreciated. Following up (a year later), I will try expand on your example to support multivariate distributions.

**Comment by jschueller** [comment]
you will need to expand the range argument and check if cdf is implemented:
```
def __init__(self, dist):
        super(ChaospyDistribution, self).__init__(len(dist))
        if len(dist)>1:
            raise Exception("Multivariate chaospy don't implement CDF computation")
        self._dist = dist
        bounds = dist.range()
        self.__range = Interval(bounds[0], bounds[1])
```

File: python/src/Distribution.i [hunk 1]
```diff
@@ -156,21 +156,27 @@ class ChaospyDistribution(PythonDistribution):
...
+        from chaospy import Iid, J, get_dependencies
+        independent = len(dist) == 1
+        independent |= isinstance(dist, Iid)
+        independent |= isinstance(dist, J) and not get_dependencies(*dist)
+        if not independent:
+            raise Exception(
+                "Dependent chaospy distributions doesn't implement CDF computation")
         self._dist = dist
         bounds = dist.range()
         self.__range = Interval(bounds[0], bounds[1])
```
Comment (jschueller) [inline python/src/Distribution.i]: you will need to extend the range attribute

Comment (jonathf) [reply python/src/Distribution.i]: I am not following.

`dist.range()` returns a numpy array of shape `(2, len(dist))`.
Looks like `Interval` takes lists as input, but accepts the numpy array of shape `(1,)` in the current implementation. I imagine that is also the case for shape `(len(dist),)`, which is what will be the case if the code is unmodified.

Do you mind explaining what about the range attribute that needs to be extended?

Comment (jschueller) [reply python/src/Distribution.i]: ok, I guess it works then

File: python/test/t_Distribution_chaospy.py [hunk 2]
```diff
@@ -37,3 +35,6 @@
     print('kurtosis=', distribution.getKurtosis())
     print('kurtosis(sampling)=', sample.computeKurtosis())
     print('range=', distribution.getRange())
+    print('scalar quantile=%.6g' % distribution.computeScalarQuantile(cdf))
```
Comment (jschueller) [inline python/test/t_Distribution_chaospy.py]: ChaospyDistribution.computeScalarQuantile should raise if dim>1

File: python/src/Distribution.i [hunk 3]
```diff
@@ -156,21 +156,27 @@ class ChaospyDistribution(PythonDistribution):
...
     --------
     >>> import openturns as ot
-    >>> # import chaospy as cp
-    >>> # chaospy_dist = st.Triangular(1.0, 2.0, 3.0)
-    >>> # distribution = ot.Distribution(ot.ChaospyDistribution(chaospy_dist))
-    >>> # distribution.getRealization()
+    >>> import chaospy as cp  # doctest: +SKIP
+    >>> chaospy_dist = cp.J(cp.Triangular(1.0, 2.0, 3.0), cp.F(4.0, 5.0))  # doctest: SKIP
+    >>> distribution = ot.Distribution(ot.ChaospyDistribution(chaospy_dist))  # doctest: SKIP
+    >>> distribution.getRealization()  # doctest: SKIP
```
Comment (jschueller) [inline python/src/Distribution.i]: isnt it "+SKIP" ? (with a +)

Comment (jonathf) [reply python/src/Distribution.i]: Yes it is. Thanks.

**Comment by jschueller** [comment]
thanks!

**Pull request merged**